In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('got_battles.csv')

In [3]:
df.head(5)

me  year  battle_number  \
0     Battle of the Golden Tooth   298              1   
1    Battle at the Mummer's Ford   298              2   
2             Battle of Riverrun   298              3   
3       Battle of the Green Fork   298              4   
4  Battle of the Whispering Wood   298              5   

              attacker_king             defender_king attacker_1 attacker_2  \
0  Joffrey/Tommen Baratheon                Robb Stark  Lannister        NaN   
1  Joffrey/Tommen Baratheon                Robb Stark  Lannister        NaN   
2  Joffrey/Tommen Baratheon                Robb Stark  Lannister        NaN   
3                Robb Stark  Joffrey/Tommen Baratheon      Stark        NaN   
4                Robb Stark  Joffrey/Tommen Baratheon      Stark      Tully   

  attacker_3 attacker_4 defender_1  ... major_death  major_capture  \
0        NaN        NaN      Tully  ...         1.0            0.0   
1        NaN        NaN  Baratheon  ...         1.0            0.0   
2        NaN        NaN      Tully  ...         0.0            1.0   
3        NaN        NaN  Lannister  ...         1.0            1.0   
4        NaN        NaN  Lannister  ...         1.0            1.0   

   attacker_size defender_size  \
0        15000.0        4000.0   
1            NaN         120.0   
2        15000.0       10000.0   
3        18000.0       20000.0   
4         1875.0        6000.0   

                                  attacker_commander  \
0                                    Jaime Lannister   
1                                     Gregor Clegane   
2                       Jaime Lannister, Andros Brax   
3  Roose Bolton, Wylis Manderly, Medger Cerwyn, H...   
4                          Robb Stark, Brynden Tully   

                                  defender_commander  summer         location  \
0                               Clement Piper, Vance     1.0     Golden Tooth   
1                                   Beric Dondarrion     1.0    Mummer's Ford   
2                      Edmure Tully, Tytos Blackwood     1.0         Riverrun   
3  Tywin Lannister, Gregor Clegane, Kevan Lannist...     1.0       Green Fork   
4                                    Jaime Lannister     1.0  Whispering Wood   

            region note  
0  The Westerlands  NaN  
1   The Riverlands  NaN  
2   The Riverlands  NaN  
3   The Riverlands  NaN  
4   The Riverlands  NaN  

[5 rows x 25 columns]

In [4]:
df['attacker_king'].value_counts()

Joffrey/Tommen Baratheon    14
Robb Stark                  10
Balon/Euron Greyjoy          7
Stannis Baratheon            5
Name: attacker_king, dtype: int64

In [5]:
df['attacker_king'].value_counts('%')

Joffrey/Tommen Baratheon    0.388889
Robb Stark                  0.277778
Balon/Euron Greyjoy         0.194444
Stannis Baratheon           0.138889
Name: attacker_king, dtype: float64

In [6]:
# Si elegimos al azar cualquier batalla
# La probabilidad que esta batalla halla sido
# Iniciada por cada atacante es de 38% por ejemplo
# Para el caso de Joffrey Baratheon
# El espacio muestreal es cada uno de los posibles elementos
# El elemento especifico es cada uno de los atacantes

In [7]:
# Espacio muestral:  Todos los posibles resultados
# Evento específico: se le asigna un número
# no negativo que representa la verosimilitud
# de ocurrencia

In [8]:
# Pr(A) = # de Ocurrencias del Evento / Total eventos totales

In [9]:
len(df)

38

In [10]:
14/38

0.3684210526315789

In [11]:
df['attacker_king'].isna().value_counts()

False    36
True      2
Name: attacker_king, dtype: int64

In [12]:
# En pandas por defecto se ignoran los datos perdidos

In [13]:
df['attackers_non_baratheon'] = np.where(df['attacker_king'] != 'Joffrey/Tommen Baratheon', 1, 0)

In [14]:
df['attackers_non_baratheon'].value_counts('%')

1    0.631579
0    0.368421
Name: attackers_non_baratheon, dtype: float64

In [15]:
batallas_rey = df['attacker_king'].value_counts('%')

In [16]:
# no negatividad

batallas_rey > 0

# Principios de funciones vectorizadas

Joffrey/Tommen Baratheon    True
Robb Stark                  True
Balon/Euron Greyjoy         True
Stannis Baratheon           True
Name: attacker_king, dtype: bool

In [17]:
# aditividad y normalización

cummulative_pr = 0
string_pr = ""

for index, value in batallas_rey.items():
    cummulative_pr += value
    string_pr += index + ""
    print(string_pr, cummulative_pr)

Joffrey/Tommen Baratheon 0.3888888888888889
Joffrey/Tommen BaratheonRobb Stark 0.6666666666666667
Joffrey/Tommen BaratheonRobb StarkBalon/Euron Greyjoy 0.8611111111111112
Joffrey/Tommen BaratheonRobb StarkBalon/Euron GreyjoyStannis Baratheon 1.0


In [18]:
df.loc[:, ['major_death', 'major_capture']].head(5)

major_death  major_capture
0          1.0            0.0
1          1.0            0.0
2          0.0            1.0
3          1.0            1.0
4          1.0            1.0

## Union de eventos y su probabilidad

In [19]:
# Para usar la union de probabilidades
# Se utiliza el operador de Python |

In [20]:
major_death_or_capture = 0
for rowname, rowserie in df.iterrows():
    if rowserie['major_death'] == 1.0 or rowserie['major_capture'] == 1.0:
        major_death_or_capture += 1
        print(rowserie['me'])

Battle of the Golden Tooth
Battle at the Mummer's Ford
Battle of Riverrun
Battle of the Green Fork
Battle of the Whispering Wood
Battle of Winterfell
Sack of Torrhen's Square
Sack of Winterfell
Battle of Oxcross
Siege of Storm's End
Sack of Harrenhal
Battle of the Blackwater
Battle of Duskendale
Retaking of Harrenhal
The Red Wedding
Siege of Seagard
Battle of Castle Black
Siege of Raventree


In [21]:
major_death_or_capture / len(df)

0.47368421052631576

In [22]:
len(df)

38

In [23]:
df['major_death_or_capture'] = np.logical_or(df['major_death'] == 1.0, df['major_capture'] == 1.0)

In [24]:
df['major_death_or_capture'].value_counts('%')

False    0.526316
True     0.473684
Name: major_death_or_capture, dtype: float64

## Intereseccion de Eventos y su probabildiad

In [26]:
major_death_and_capture = 0

for rowname, rowserie in df.iterrows():
    if rowserie['major_death'] == 1.0 and rowserie['major_capture'] == 1.0:
        print(rowserie['me'])
        major_death_and_capture += 1


Battle of the Green Fork
Battle of the Whispering Wood
Battle of Oxcross
Battle of the Blackwater
The Red Wedding
Battle of Castle Black


In [27]:
major_death_and_capture / len(df)

0.15789473684210525

La probabilidad de ambos eventos es del 16%

Se puede usar un logical and

In [31]:
df['major_death_and_capture'] = np.logical_and(df['major_death'] == 1.0,
               df['major_capture'] == 1.0)

In [32]:
df['major_death_and_capture'].value_counts('%')

False    0.842105
True     0.157895
Name: major_death_and_capture, dtype: float64

Obtenemos el mismo resultado

## Probabilidad Condicional

El objetivo es responder a la verosimilidud de ocurrencia
de un evento A, condicional a la ocurrencia en otro evento B

$Pr\left( A | B \right) =\frac{Pr \left( A \bigcap B \right) }{Pr \left( B \right) }$

**Nota**:  Para utilizar notación matemática en el markdown, se puede ver en: https://en.wikibooks.org/wiki/LaTeX/Mathematics

In [68]:
muerte_importante_y_baratheon = 0

for rowname, rowserie in df.iterrows():
    if rowserie['major_death'] == 1.0 and rowserie['attacker_king'] == "Joffrey/Tommen Baratheon":
            print(rowserie['me'])
            muerte_importante_y_baratheon += 1

print(muerte_importante_y_baratheon)

Battle of the Golden Tooth
Battle at the Mummer's Ford
Sack of Winterfell
Retaking of Harrenhal
The Red Wedding
5


In [41]:
batalla_baratheon = df['attacker_king'].value_counts().get('Joffrey/Tommen Baratheon')
print(batalla_baratheon)

14


In [42]:
muerte_importante_y_baratheon / batalla_baratheon

0.35714285714285715

Hay un 36 porciento de probabilidad de que se produzca una muerte
importante cuando se inicia una batalla por un Baratheon

In [57]:
df['major_death_and_baratheon'] = np.logical_and(df['major_death'] == 1.0,
                                                  df['attacker_king'] == "Joffrey/Tommen Baratheon")

In [93]:
df['major_death_and_baratheon'].value_counts()

False    33
True      5
Name: major_death_and_baratheon, dtype: int64

In [94]:
df['major_death_and_baratheon'].value_counts('%')

False    0.868421
True     0.131579
Name: major_death_and_baratheon, dtype: float64

In [90]:
major_death_and_baratheon = df['major_death_and_baratheon'].value_counts().get(True)
print(major_death_and_baratheon)

5


In [92]:
major_death_and_baratheon_prob = df['major_death_and_baratheon'].value_counts('%').get(True)
print(major_death_and_baratheon_prob)

0.13157894736842105


In [97]:
major_death_and_baratheon_prob == 5/len(df)

True

In [95]:
df['attacker_king'].value_counts()

Joffrey/Tommen Baratheon    14
Robb Stark                  10
Balon/Euron Greyjoy          7
Stannis Baratheon            5
Name: attacker_king, dtype: int64

In [96]:
df['attacker_king'].value_counts('%')

Joffrey/Tommen Baratheon    0.388889
Robb Stark                  0.277778
Balon/Euron Greyjoy         0.194444
Stannis Baratheon           0.138889
Name: attacker_king, dtype: float64

In [98]:
batalla_baratheon_prob = df['attacker_king'].value_counts('%').get('Joffrey/Tommen Baratheon')
print(batalla_baratheon_prob)

0.3888888888888889


In [99]:
batalla_baratheon_prob == 14/len(df)

False

In [81]:
major_death_and_baratheon_prob / batalla_baratheon_prob

0.3383458646616541

In [101]:
major_death_and_baratheon_prob / (14/len(df))

0.35714285714285715

Cúal es la probabilidad correcta?  Es mejor utilizar 5/14: 0.35714285714285715

# Funciones

In [115]:
def hipotenusa(cateto1=3, cateto2=4):
    """
    Definicion: Qué realiza esta función?
    Devuelve la hipotenusa en base a dos catetos entregados por el usuario
    Parámetros de ingreso:  Qué entra?
    - cateto1: Tipo de dato: int.  Valor del primer cateto a ingresar que será
    elevado al cuadrado posteriormente.  Por defecto tiene el valor 3
    - cateto2: Tipo de dato: int.  Valor del segundo cateto a ingresar que será
    elevado al cuadrado posteriormente.  Por defecto tiene el valor 4
    Retorno: Qué sale?
    - Retorna la hipotenusa. Por defecto me devuelve un float.
    """
    return np.sqrt(cateto1 ** 2 + cateto2 ** 2)

In [112]:
hipotenusa(6, 8)

10.0

In [113]:
hipotenusa()

5.0

## Primera función: Reporte de estadísticos univariados

Objetivo:  Extraer la media, mediana y varianza de un ```pd.Series```

In [119]:
from scipy import stats

def fetch_descriptives(x):
    tmp_mean = np.mean(x)
    tmp_median = np.median(x)
    tmp_var = np.var(x)
    tmp_mode = stats.mode(x)
    
    return tmp_mean, tmp_median, tmp_mode, tmp_var

In [121]:
fetch_descriptives(df['attacker_size'])

/Users/luisalfredoporraspaez/anaconda3/lib/python3.7/site-packages/numpy/lib/function_base.py:3405: RuntimeWarning: Invalid value encountered in median
  r = func(a, **kwargs)


(9942.541666666666,
 nan,
 ModeResult(mode=array([6000.]), count=array([3])),
 394261997.7482639)

La mediana está devolviendo nan, debe ser que hay un dato que es nan
entonces debemos proceder a limpiar la data antes de darsela a la función

In [129]:
from scipy import stats

def fetch_descriptives(x):
    x = x.dropna()
    tmp_mean = np.mean(x)
    tmp_median = np.median(x)
    tmp_var = np.var(x)
    tmp_mode = stats.mode(x)[0][0]
    
    return tmp_mean, tmp_median, tmp_mode, tmp_var

In [130]:
fetch_descriptives(df['attacker_size'])

(9942.541666666666, 4000.0, 6000.0, 394261997.7482639)

In [124]:
stats.mode(df['attacker_size'])

ModeResult(mode=array([6000.]), count=array([3]))

In [128]:
stats.mode(df['attacker_size'])[0][0]

6000.0

In [133]:
descriptives_attack_size = fetch_descriptives(df['attacker_size'])
descriptives_attack_size

(9942.541666666666, 4000.0, 6000.0, 394261997.7482639)

In [134]:
mean_attacker, median_attacker, mode_attacker, var_attacker = fetch_descriptives(df['attacker_size'])

In [135]:
fetch_descriptives(df)

/Users/luisalfredoporraspaez/anaconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:3118: RuntimeWarning: Mean of empty slice.
  out=out, **kwargs)


ZeroDivisionError: division by zero

Para evitar este error vamos a incluir validaciones.

In [136]:
isinstance(df['attacker_size'], pd.Series)

True

In [137]:
type(df['attacker_size'])

pandas.core.series.Series

In [141]:
from scipy import stats

def fetch_descriptives(x):
    if isinstance(x, pd.Series) is True:
        if x.dtype != 'object':
            x = x.dropna()
            tmp_mean = np.mean(x)
            tmp_median = np.median(x)
            tmp_var = np.var(x)
            tmp_mode = stats.mode(x)[0][0]
        else:
            raise TypeError('La serie no contiene datos del tipo numérico')
    else:
        raise TypeError('El dato ingresado no es una serie.')
    
    return tmp_mean, tmp_median, tmp_mode, tmp_var

In [142]:
fetch_descriptives(df)

TypeError: El dato ingresado no es una serie.

In [152]:
fetch_descriptives(df['attacker_king'])

TypeError: La serie no contiene datos del tipo numérico

In [153]:
df['attacker_king'].dtype

dtype('O')

In [156]:
df['attacker_king'].dtype == 'object'

True

In [157]:
str(df['attacker_king'].dtype)

'object'

## Segunda función: Cálculo de probabilidad condicional

Objetivo: dado un ```pd.DataFrame``` y dos ```pd.Series```, evaluar la propabilidad de $Pr(A|B)$ donde $A$ y $B$ son eventos ingresados por el usuario.

Fórmula a Implementar:

$Pr\left( A | B \right) =\frac{Pr \left( A \bigcap B \right) }{Pr \left( B \right) }$

In [199]:
def get_conditional_pr(df, A_var, A_event, B_var, B_event):
    
    tmp_intersection_counter = 0
    # Verificar si es que las variables existen en el dataframe
    if A_var in df and B_var in df:
    
        # verificar si es que los eventos existen en las variables declaradas
        if A_event in df[A_var].values and B_event in df[B_var].values:
            # generar el mismo resultado pero con numpy.
            #for _, rowserie in df.iterrows():
            #    if rowserie[A_var] == A_event and rowserie[B_var] == B_event:
            #      tmp_intersection_counter += 1
            
            tmp_intersection_counter = np.logical_and(df[A_var] == A_event,
                           df[B_var] == B_event).value_counts().get(True)
            tmp_b_event_freq = df[B_var].value_counts().get(B_event)

            tmp_result = tmp_intersection_counter / tmp_b_event_freq

            print(f"Pr({A_event} | {B_event}) = {tmp_result}")

            return tmp_result
        else:
            raise TypeError("Los eventos no existen en las variables declaradas")
    
    else:
        raise TypeError('Las variables no existen en el DataFrame')

In [200]:
get_conditional_pr(df, A_var='major_death', A_event=1.0,
                   B_var='attacker_king', B_event='Joffrey/Tommen Baratheon')

Pr(1.0 | Joffrey/Tommen Baratheon) = 0.35714285714285715


0.35714285714285715

In [186]:
len(df[df['major_death'] == 1.0])

13

In [187]:
1.0 in df['major_death'].values

True

In [202]:
100 + 23 + 67 + 87 + 91 + 56 + 101 + 34

559

In [203]:
23/559

0.04114490161001789

In [205]:
100 + 101 + 34

235

In [206]:
235 / 559

0.4203935599284436

In [207]:
34 + 26 + 15

75

In [208]:
34 / 75

0.4533333333333333

In [209]:
34 + 20 + 6

60

In [210]:
(20 + 6) / 60

0.43333333333333335